src/lua/log.lua

```diff
@@ -636,6 +645,123 @@ local function load_cfg(self, cfg)
     say(S_DEBUG, m:format(cfg.log, cfg.level, cfg.nonblock, cfg.format))
 end
 
+local option_types = {
+    log = 'string',
+    nonblock = 'boolean',
+    level = 'number, string',
+    format = 'string',
+}
+
+local log_initialized = false
+
+local function log_C_cfg(cfg)
+    local cfg_C = table.copy(cfg)
+    if type(cfg.level) == 'string' then
+        cfg_C.level = log_level_keys[cfg.level]
+    end
+    local nonblock
+    if cfg.nonblock ~= nil then
+        nonblock = cfg.nonblock and 1 or 0
+    else
+        nonblock = -1
+    end
+    cfg_C.nonblock = nonblock
+    return cfg_C
+end
+
+local function log_check_cfg(cfg)
+    if type(cfg.level) == 'string' and
+       log_level_keys[cfg.level] == nil then
+        local err = ("expected %s"):format(log_level_list())
+        box.error(box.error.CFG, "log_level", err)
+    end
+
+    if log_initialized then
+        if log_cfg.log ~= cfg.log then
+            box.error(box.error.RELOAD_CFG, 'log');
+        end
+        if log_cfg.nonblock ~= cfg.nonblock then
+            box.error(box.error.RELOAD_CFG, 'log_nonblock');
+        end
+    end
+
+    local cfg_C = log_C_cfg(cfg)
+    if ffi.C.say_check_cfg(cfg_C.log, cfg_C.level,
+                           cfg_C.nonblock, cfg_C.format) ~= 0 then
+        box.error()
+    end
+end
+
+local function box_cfg_update(key)
+    if key == nil then
+        for km, kb  in pairs(log2box_keys) do
+            box.internal.cfg[kb] = log_cfg[km]
+        end
+    else
+        box.internal.cfg[log2box_keys[key]] = log_cfg[key]
+    end
+end
+
+local function set_log_level(level)
+    box.internal.check_cfg_option_type(option_types.level, 'level', level)
+    local cfg = table.copy(log_cfg)
+    cfg.level = level
+    log_check_cfg(cfg)
+
+    local cfg_C = log_C_cfg(cfg)
+    ffi.C.say_set_log_level(cfg_C.level)
+    log_cfg.level = level
+
+    box_cfg_update('level')
+
+    local m = "log: level set to %s"
+    log.debug(m:format(level))
+end
+
+local function set_log_format(format)
+    box.internal.check_cfg_option_type(option_types.format, 'format', format)
+    local cfg = table.copy(log_cfg)
+    cfg.format = format
+    log_check_cfg(cfg)
+
+    ffi.C.say_set_log_format(fmt_str2num[format])
+    log_cfg.format = format
+
+    box_cfg_update('format')
+
+    local m = "log: format set to '%s'"
+    log.debug(m:format(format))
+end
+
+local function log_configure(self, cfg)
+    cfg = box.internal.prepare_cfg(cfg, default_cfg, option_types)
+    box.internal.merge_cfg(cfg, log_cfg);
```

Comment:
Looks like `prepare_cfg` and `merge_cfg` are always used together (in `load_cfg.lua` too). Maybe, merge them to make the internal API simpler?